pkg/network/protocols/http/shared_libraries.go

```diff
@@ -160,221 +257,58 @@ func (w *soWatcher) Start() {
+	r.m.Lock()
+	defer r.m.Unlock()
 
-		if len(registration.pids) == 0 {
-			// all processes are gone, so schedule this for removal
-			pathsToBeRemoved[libpath] = struct{}{}
-			regsToBeRemoved[registration] = struct{}{}
-			updated = true
-		}
+	reg, found := r.byPID[pid]
+	if !found {
+		return
 	}
-
-	for libpath := range pathsToBeRemoved {
-		w.registry.unregister(libpath)
+	if reg.Unregister() == true {
+		// we need to cleanup our entries as there are no more process using this ELF
+		delete(r.byID, reg.pathID)
 	}
-
-	return updated
+	delete(r.byPID, pid)
 }
 
-func (w *soWatcher) canonicalizePath(path string) string {
-	if w.hostMount != "" {
-		path = filepath.Join(w.hostMount, path)
-	}
-
-	return followSymlink(path)
-}
-
-type soRegistration struct {
-	// set of pid paths (eg. /host/proc/123) that are used for liveness checks
-	// this is done to unregister shared libraries as soon as all processes
-	// referencing it terminate. This is required when the shared library is
-	// located in a volume where mount propagation is enabled, so containers
-	// don't get stuck during termination.
-	pids map[string]struct{}
-
-	inode        uint64
-	refcount     int
-	unregisterCB func(string) error
-}
-
-type soRegistry struct {
-	procRoot string
-	byPath   map[string]*soRegistration
-	byInode  map[uint64]*soRegistration
-}
-
-func (r *soRegistry) register(libPath string, pid int, rule soRule) {
-	pidPath := fmt.Sprintf("%s/%d", r.procRoot, pid)
-	if registration, ok := r.byPath[libPath]; ok {
-		registration.pids[pidPath] = struct{}{}
+func (r *soRegistry) Register(root string, libPath string, pid uint32, rule soRule) {
+	hostLibPath := root + libPath
+	pathID, err := newPathIdentifier(hostLibPath)
+	if err != nil {
+		// short living process can't hit here
+		// as we receive the openat() syscall info after receiving the EXIT netlink process
+		log.Debugf("can't create path identifier %s", err)
 		return
 	}
-
-	inode, err := getInode(libPath)
-	if err != nil {
+	if _, found := r.blocklistByID[pathID]; found {
 		return
 	}
 
-	if registration, ok := r.byInode[inode]; ok {
-		registration.refcount++
-		r.byPath[libPath] = registration
-		registration.pids[pidPath] = struct{}{}
-		log.Debugf("registering library=%s", libPath)
+	r.m.Lock()
+	defer r.m.Unlock()
+	if reg, found := r.byID[pathID]; found {
+		reg.refcount++
+		r.byPID[pid] = reg
 		return
 	}
 
-	err = rule.registerCB(libPath)
-	if err != nil {
-		log.Debugf("error registering library=%s: %s", libPath, err)
-		if err := rule.unregisterCB(libPath); err != nil {
-			log.Debugf("unregisterCB %s : %s", libPath, err)
+	if err := rule.registerCB(pathID, root, libPath); err != nil {
+		log.Debugf("error registering library %s path %s by pid %d : %s", pathID.String(), hostLibPath, pid, err)
+		if rule.unregisterCB != nil {
+			if err := rule.unregisterCB(pathID); err != nil {
+				log.Debugf("unregisterCB library %s path %s : %s", pathID.String(), hostLibPath, err)
+			}
 		}
```

Comment:
why do we need to unregister a callback that failed to register?

Reply:
the old implementation use this mechanism of cleanup callback
As the register callback can failed to register any uprobe function, and they are registered one at a time in the kernel
Calling the unregister on error would cleanup already registered functions

Reply:
I added a comment